File: Interpreter/CalculateFunction.java
package calculator.xwg;

import java.util.LinkedList;

public abstract class CalculateFunction {
	CalculateFunction(){
		
	}
	
	boolean checkResult(EvaluateContext context){
		if(context.getCurrentResult().r == Double.NaN){
			context.setErrorMessage(getName(), R.string.error_mathmatic_error);
			return false;
		}
		return true;
    }
    String getName(){ return ""; }
   
    boolean execute(LinkedList<Complex> paraList, EvaluateContext context){ return false; }
}

File: Interpreter/CalculateFunction.py
import abc
import math
from Interpreter.R_string import *

class CalculateFunction(metaclass=abc.ABCMeta):
    def checkResult(self, context):
        if math.isnan(context.getCurrentResult().r):
            context.setErrorMessage(self.get_name(), R_string.error_mathmatic_error)
            return False
        return True
    @abc.abstractmethod
    def get_name(self):
        return ''

    @abc.abstractmethod
    def execute(self, context, paraList):
        return False
